### ISSUE TITLE
Nested models

### ISSUE THREAD

**Issue by Jejulia** [JuliaStats/MixedModels.jl#445]
Hi:
I was fitting models with nested grouping factors, like `y ~ x + (1|A) + (1|A&B)`, but errors occurred:
```
ERROR: LoadError: MethodError: no method matching strides(::Diagonal{Float64,Array{Float64,1}})  
...
Stacktrace:
 [1] stride(::Diagonal{Float64,Array{Float64,1}}, ::Int64) at .\abstractarray.jl:396
 [2] stride1(::Diagonal{Float64,Array{Float64,1}}) at C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\LinearAlgebra\src\LinearAlgebra.jl:197
 [3] _chkstride1 at C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\LinearAlgebra\src\LinearAlgebra.jl:203 [inlined]
 [4] chkstride1 at C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\LinearAlgebra\src\LinearAlgebra.jl:201 [inlined]
 [5] syrk!(::Char, ::Char, ::Float64, ::Array{Float64,2}, ::Float64, ::Diagonal{Float64,Array{Float64,1}}) at C:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\LinearAlgebra\src\blas.jl:1594
 [6] rankUpdate!(::Hermitian{Float64,Diagonal{Float64,Array{Float64,1}}}, ::Array{Float64,2}, ::Float64, ::Float64) at C:\Users\sciph\.julia\packages\MixedModels\36RJL\src\linalg\rankUpdate.jl:24
...
```
I've tried different datasets (which worked in `lme4` or `nlme`), but same errors kept occuring. 

It seems that the problem is using `BLAS.syrk!` which will eventually stride the diagonal matrix.
Is there an alternative way to avoid `Diagonal` or other suggestions? Thanks!

**Comment by palday** [comment]
If the problem is fundamentally at `strides`, then it's a `LinearAlgebra` and not `MixedModels` issue. But first we need more information to see if something is going wrong before then:

- What Julia version? (looks like 1.5 from the stack trace; which patch version?)
- What MixedModels version?
- Can you share a minimum working example or (even privately) your data so that we can test and debug this issue?

**Comment by Jejulia** [comment]
I am using Julia 1.5.3 and `MixedModels@3.1.1`
I was testing my other package which requiring `MixedModels` with public data.
The first data I used was `Oxide` from `nlme`. I fit the model as `Thickness ~ 1 + (1|Lot) + (1|Lot&Wafer)`. 
The second data I used was from [http://bayes.acs.unt.edu:8083/BayesContent/class/Jon/R_SC/Module9/lmm.data.txt](url). I read the data with header and set the sep as ",".
The model was `extro ~ open + agree + social + (1|school) + (1|school&class)`
I've tried both data in R which worked. Thanks for help!

**Comment by palday** [comment]
@dmbates I suspect we might be able to sidestep the issue by taking adding a `rankUpdate!(::Diagonal, ::Diagonal)` method, which would probably be worthwhile as a particularly fast special case, but I haven't gotten that far yet:
```julia
julia> BlockDescription(m)
rows: Lot & Wafer       Lot          fixed     
  24:   Diagonal   
   8:    Dense        Diagonal   
   1:    Dense         Dense         Dense     
```

**Comment by palday** [comment]
We have a fix, which you can install via

```julia
julia>] add MixedModels#pa/diagonalstride
```

It works for the Oxide data. Can you check your other tests? If all goes well, we'll get this merged in and a new patch release in a a day or so. (We have a few other compatibility updates so it's good time for another patch release anyway.)

**Comment by palday** [comment]
The patch release is is already scheduled and should be out tonight.  If you find that things still don't work, please feel free to reopen the issue.

**Comment by Jejulia** [comment]
Thanks! I've tested the second one and it worked, but then I tried random slope and error occurred again with a similar problem.
I used `Oxide` and fit it like `Thickness ~ Source + (Source|Lot) + (Source|Lot&Wafer)`
```
ERROR: LoadError: MethodError: no method matching strides(::UniformBlockDiagonal{Float64})
```
It's the same function but different type. I guess another specific method for `BlockDiagonal` is needed.

**Comment by palday** [comment]
@Jejulia The next fix is already on its way in #447. :smile: Note that the the `Oxide` with random slopes everywhere seems to have a poorly defined optimum -- different optimizers in lme4 give somewhat different answers. Most of those differences seem to be in the correlations and not the standard deviations, and the fixed effects and their standard errors seem to be stable across optimizers, but I would still be careful about depending too much on that model+dataset as an example or test.

**Comment by palday** [comment]
And yet another patch release is out. If you run into more missing methods, do let us know.
